plugins/usbdmx/AsyncPluginImpl.cpp

```diff
@@ -353,17 +346,21 @@ bool AsyncPluginImpl::StartAndRegisterDevice(Widget *widget, Device *device) {
 
 /*
  * @brief Signal widget removal.
- * @param device_id The device id to remove.
+ * @param state The DeviceState owning the widget to be removed.
  *
- * This is run within the main thread.
+ * This is run within the main thread and it must. The Device destructor
+ * below may cause libusb_close() to be called, which will deadlock if the
+ * hotplug event thread waits for it.
  */
-void AsyncPluginImpl::ShutdownDevice(Device *device, Future<void> *f) {
-  m_plugin_adaptor->UnregisterDevice(device);
-  device->Stop();
-  delete device;
-  if (f) {
-    f->Set();
+void AsyncPluginImpl::ShutdownDeviceState(DeviceState *state) {
+  if (state->ola_device) {
+    Device *device = state->ola_device;
+    m_plugin_adaptor->UnregisterDevice(device);
+    device->Stop();
+    delete device;
+    state->ola_device = NULL;
   }
```

Comment:
It would be potentially nice to log at say WARN or DEBUG level if this happens with a state without ola_device, which presumably means its getting called twice or something? With a log level appropriate to whether it can happen via legitimate means (e.g. if you pull a device OLA doesn't care about).

Reply:
You mean spit out a warning if ola_device is already NULL, right?

IIRC ola_device being NULL here is legitimate because of the way that StartAandRegisterDevice is called by the factory through BaseWidgetFactory::AddWidget and AsyncPluginImpl::NewWidget. The Factory can actually delay calling AddWidget instead of doing it right away in the DeviceAdded function like normal. If the usb device is then unplugged before it does end up calling AddWidget then ola_device will be NULL here.

Reply:
> You mean spit out a warning if ola_device is already NULL, right?

Yeah

> IIRC ola_device being NULL here is legitimate because of the way that StartAandRegisterDevice is called by the factory through BaseWidgetFactory::AddWidget and AsyncPluginImpl::NewWidget. The Factory can actually delay calling AddWidget instead of doing it right away in the DeviceAdded function like normal. If the usb device is then unplugged before it does end up calling AddWidget then ola_device will be NULL here.

Makes sense. It might be worth adding one as OLA_DEBUG then; it sounds like it should only happen in a few rare circumstances, so being able to see it in a log could be handy if its firing at other times.

Reply:
I've added the debug output and a comment explaining the situation when this might legitimately arise.

Reply:
Thanks.